### PR TITLE
Initial draft of the function-builders proposal

### PR DESCRIPTION
**Pull request by rjmccall**

### CODE REVIEW

File: proposals/XXXX-function-builders.md [hunk 1]
```diff
@@ -0,0 +1,789 @@
...
+  div {
+    ...
+  }
+  div {
+    ...
+  }
+}
+```
+
+The DSL has to be embedded into the ordinary language somehow, which means that at least the outermost layer must obey something like ordinary language rules.  Under ordinary language rules, this is a function call to `body `passing a trailing closure.  It makes sense, then, that what we're doing is taking the body of the anonymous function and apply some sort of transformation to it.  This raises a series of separate questions:
```
Comment (AliSoftware) [inline proposals/XXXX-function-builders.md]:
```suggestion
The DSL has to be embedded into the ordinary language somehow, which means that at least the outermost layer must obey something like ordinary language rules.  Under ordinary language rules, this is a function call to `body` passing a trailing closure.  It makes sense, then, that what we're doing is taking the body of the anonymous function and apply some sort of transformation to it.  This raises a series of separate questions:
```

**Comment by rjmccall** [comment]
> Just small nits / typo fixes

Thanks, applied!

File: proposals/XXXX-function-builders.md [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+
+These last two points (and some other considerations) strongly suggest that the DSL should be identified with a type that can provide an arbitrary namespace of functions that can be used as *ad hoc* customization points for the transformation.
+
+## Detailed design
+
+[SE-0258](https://github.com/apple/swift-evolution/blob/master/proposals/0258-property-delegates.md) introduced the concept of custom attributes to Swift, and while that proposal is still under development on its main subject, the Core Team has accepted the portions relating to custom attributes as a useful technical direction.  We will build on that here.
```
Comment (zntfdr) [inline proposals/XXXX-function-builders.md]: SE-0258 has been renamed to `0258-property-wrappers.md`, therefore the link should be updated to  `https://github.com/apple/swift-evolution/blob/master/proposals/0258-property-wrappers.md` (currently it displays the GitHub 404 page) 😊

**Comment by DougGregor** [comment]
This is going to be heavily revised; we can open a new PR at the point where it is ready.

**Pull request closed without merging**